ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+    promise_2.result # => 7
+
+    # This would block until the promise is resolved (longer delay)
+    promise_1.result # => 3
+  end
+
+  private
+
+  def wait_and_sum(a, b, delay:)
+    workflow.sleep(delay)
+    workflow.execute_activity(SumActivity, a, b)
+  end
+end
+```
+
+Methods `async.all` and `async.any` return promises themselves, so they can be used just like the
+promises returned by an `async {}` call.
+
+Also worth noting that `async {}` calls can be nested to model more complex business logic.
+
+
+## Activity heartbeats & cancellations
+
+This approach uses `Thread#raise` that will eject the execution out of the current context. In cases
+where this is not acceptable we will allow to either wrap a critical code in an `activity.shield`
+block (that will not raise until the block has executed) or opt-out of this default behaviour
+altogether and use `activity.cancelled?` flag to determine when an activity was cancelled.
+
+```ruby
+class LongRunningActivity < Temporal::Activity
+  def execute
+    10.times do
+      sleep 1
+      activity.heartbeat('ping')
+    end
+  rescue Temporal::Error::ActivityCancelled
+    raise 'this activity is cancelled'
+  end
+end
+```
+
+Here is what it will look like from a workflow perspective:
+
+```ruby
+class InnerCancelWorkflow < Temporal::Workflow
+  def execute
+    # the block supplies an optional argument
+    async do |outer_promise|
+      async do
+        workflow.sleep(30)
+        outer_promise.cancel
+      end
+
+      # One of these blocking activities will end up getting cancelled 
+      #   unless they all finish within 30 seconds
+      workflow.execute_activity(LongRunningActivity)
+      workflow.execute_activity(LongRunningActivity)
+      workflow.execute_activity(LongRunningActivity)
+    end.await
+  end
+end
+```
+
+#### Alternative approaches
+
+- Overall using `Thread#raise` is considered bad practice because it might raise in the
+  middle of some process. We might want to consider opting in to this behaviour instead
+
+
+## Workflow cancellations
+
+To handle external workflow cancellations (requested by a client or another workflow) we will use a
+similar approach to activities. By default the cancellation will attempt to cancel all currently
+executing fibers. To protect critical functionality from cancellations a `workflow.shield` method
+can be used. In which case cancellation will raise after the block is executed:
+
+
+```ruby
+class MyWorkflow < Temporal::Activity
+  def execute
+    # Can get cancelled
+    workflow.execute_activity(NonCriticalActivity)
+
+    # This block will delay any cancellations until it is finished
+    async(shield: true) do
+      workflow.execute_activity(CriticalActivity)
+      workflow.execute_activity(OtherCriticalActivity)
+    end.await
+    
+    # Can get cancelled
+    workflow.execute_activity(AnotherNonCriticalActivity)
+  rescue Temporal::Error::WorkflowCancelled
+    ...
```

Comment:
Do you expect that more activities can be run here as well? I do like that creating a canceled context does not seem to be necessary. I've seen that abstraction be confusing to workflow authors.

Reply:
> Do you expect that more activities can be run here as well?

Yes, this is a feature in all Temporal SDKs. But note, cancellation is a reported state not an event from the Temporal server. So you can't get a second cancellation. And if this error is not bubbled out, the workflow won't be considered cancelled.